File: Data_Structures/leetcode-blind75/maxSubArray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let maxSub = nums[0];//there is always going to be a num 
    let currentSum = 0;
    
    for (let i = 0; i < nums.length; i++) {
        //we are going to remove negative prefix as we computed the         totally sum 
        if (currentSum < 0) {
            currentSum = 0;
        }
        currentSum += nums[i];
        //updating the maxSub
        maxSub = Math.max(maxSub, currentSum)
    }
    return maxSub;
};

File: leetcode/MaxSubArray.cs
public class Solution {
    public int MaxSubArray(int[] nums) {
        int maxSum = nums[0];
        int currentSum = 0;
        
        for (int i = 0; i < nums.Length; i++) 
        {
            if (currentSum < 0)
            {
                currentSum = 0;
            }
            
            currentSum += nums[i];
            maxSum = Math.Max(maxSum, currentSum);
        }
        return maxSum;
    }
}